### PR TITLE
[DOC] Add join processor documentation

### PR DESCRIPTION
**Pull request by vagimeli**
### Description
Closes content gap

### Issues Resolved
#4647 

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by vagimeli** [comment]
@hdhalter We need a dev to take the first pass at drafting this content. I can support them in refining the content and getting it through the documentation process. The ingest processor template is provided with this PR, so it should help the dev get started. Thanks!

File: _ingest-pipelines/processors/join.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+### Step 2 (Optional): Test the pipeline
+
+It is recommended that you test your pipeline before you ingest documents.
+{: .tip}
+
+To test the pipeline, run the following query:
+
+```json
+<insert code example>
+```
```
Comment (vagimeli) [inline _ingest-pipelines/processors/join.md:49]: @gaobinlong Could you help me troubleshoot this pipleline example? I'm getting the following error: {
  "docs": [
    {
      "error": {
        "root_cause": [
          {
            "type": "illegal_argument_exception",
            "reason": "field [message] of type [java.lang.String] cannot be cast to [java.util.List]"
          }
        ],
        "type": "illegal_argument_exception",
        "reason": "field [message] of type [java.lang.String] cannot be cast to [java.util.List]"
      }
    }
  ]
}

File: _ingest-pipelines/processors/join.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+---
+layout: default
+title: Join
+parent: Ingest processors
+nav_order: 160
+---
+
+# Join processor
+
+The `join` processor combines fields from different data sources into a single document before indexing. Raises an error if the field is not an array. For example, you could combine the log message, severity level, and timestamp into a single field for better readability and easier querying. Or, you could join data from different sources, such as application logs and system logs, based on a common field like a session ID or user ID to provide a more comprehensive view of related events and help in troubleshooting and root cause analysis.
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:10]: `join` processor can only be used to join every elements in an array into a single string value, join data from different sources seems not possible, see the [comment](https://github.com/opensearch-project/OpenSearch/blob/27003270adedebd43eedbd0186c0825f2c0e8613/modules/ingest-common/src/main/java/org/opensearch/ingest/common/JoinProcessor.java#L45) in the code.

File: _ingest-pipelines/processors/join.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+{% include copy-curl.html %}
+
+## Configuration parameters
+
+The following table lists the required and optional parameters for the `join` processor.
+
+Parameter | Required/Optional | Description |
+|-----------|-----------|-----------|
+`field` | Required | The field name where the join operator is applied. 
+`separator` | Optional | A string separator to use when joining field values. If not specified, the values are concatenated without a separator.
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:31]: `separator` is required.

File: _ingest-pipelines/processors/join.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+```
+{% include copy-curl.html %}
+
+## Configuration parameters
+
+The following table lists the required and optional parameters for the `join` processor.
+
+Parameter | Required/Optional | Description |
+|-----------|-----------|-----------|
+`field` | Required | The field name where the join operator is applied.
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:30]: `field` must be an array, we should emphasize this.

File: _ingest-pipelines/processors/join.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+## Using the processor
+
+Follow these steps to use the processor in a pipeline.
+
+### Step 1: Create a pipeline
+
+The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 
+
+```json
+PUT _ingest/pipeline/example-join-pipeline
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:48]:
```
PUT _ingest/pipeline/example-join-pipeline
{
  "description": "Example pipeline using the join processor",
  "processors": [
    {
      "join": {
        "field": "uri",
        "separator": "/"
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/join.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+### Step 2 (Optional): Test the pipeline
+
+It is recommended that you test your pipeline before you ingest documents.
+{: .tip}
+
+To test the pipeline, run the following query:
+
+```json
+POST _ingest/pipeline/example-join-pipeline/_simulate
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:72]:
```
POST _ingest/pipeline/example-join-pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "uri": [
          "app",
          "home",
          "overview"
        ]
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/join.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+}
+```
+{% include copy-curl.html %}
+
+#### Response
+
+The following example response confirms that the pipeline is working as expected:
+
+```json
+
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:92]:
```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "uri": "app/home/overview"
        },
        "_ingest": {
          "timestamp": "2024-05-24T02:16:01.00659117Z"
        }
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/join.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+### Step 3: Ingest a document 
+
+The following query ingests a document into an index named `testindex1`:
+
+```json
+<insert code example>
+```
+{% include copy-curl.html %}
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:102]:
```
POST testindex1/_doc/1?pipeline=example-join-pipeline
{
  "uri": [
    "app",
    "home",
    "overview"
  ]
}
```

File: _ingest-pipelines/processors/join.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+```json
+<insert code example>
+```
+
+### Step 4 (Optional): Retrieve the document
+
+To retrieve the document, run the following query:
+
+```json
+<insert code example>
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:117]: `GET testindex1/_doc/1`

File: _ingest-pipelines/processors/join.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+### Step 3: Ingest a document 
+
+The following query ingests a document into an index named `testindex1`:
+
+```json
+<insert code example>
+```
+{% include copy-curl.html %}
+
+#### Response
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/join.md:104]: Seems this section is not needed.

File: _ingest-pipelines/processors/join.md (hunk 2 above)
Comment (vagimeli) [reply _ingest-pipelines/processors/join.md]: Revised.

File: _ingest-pipelines/processors/join.md (hunk 4 above)
Comment (vagimeli) [reply _ingest-pipelines/processors/join.md]: Revised

File: _ingest-pipelines/processors/join.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+`on_failure` | Optional | Handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
+`tag` | Optional | Identifier for the processor. Useful for debugging and metrics.
+
+## Using the processor
+
+Follow these steps to use the processor in a pipeline.
+
+### Step 1: Create a pipeline
+
+The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>:
```
Comment (vagimeli) [inline _ingest-pipelines/processors/join.md:45]:
```suggestion
The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 
```
```suggestion
The following query creates a pipeline named `example-join-pipeline` that uses the `join` processor to concatenate all the values of the `uri`  field, separating them with the specified separator `/`: 
```

**Comment by vagimeli** [comment]
@gaobinlong Thank you for your review and the examples. I've added your changes.

File: _ingest-pipelines/processors/join.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+```
+{% include copy-curl.html %}
+
+## Configuration parameters
+
+The following table lists the required and optional parameters for the `join` processor.
+
+Parameter | Required/Optional | Description |
+|-----------|-----------|-----------|
+`field` | Required | The field name where the join operator is applied. Must be an array.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:30]:
```suggestion
`field` | Required | The name of the field to which the join operator is applied. Must be an array.
```

File: _ingest-pipelines/processors/join.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+`separator` | Required | A string separator to use when joining field values. If not specified, the values are concatenated without a separator.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:31]:
```suggestion
`separator` | Required | A string separator to use when joining field values. If not specified, then the values are concatenated without a separator.
```

File: _ingest-pipelines/processors/join.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+`target_field` | Optional | The field to assign the cleaned value to. If not specified, field is updated in-place.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:32]:
```suggestion
`target_field` | Optional | The field to assign the cleaned value to. If not specified, then the field is updated in place.
```

File: _ingest-pipelines/processors/join.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+`description` | Optional | Description of the processor's purpose or configuration.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:33]:
```suggestion
`description` | Optional | A description of the processor's purpose or configuration.
```

File: _ingest-pipelines/processors/join.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+`if` | Optional | Conditionally execute the processor.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:34]:
```suggestion
`if` | Optional | Specifies to conditionally execute the processor.
```

File: _ingest-pipelines/processors/join.md [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+`ignore_failure` | Optional | Ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:35]:
```suggestion
`ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

File: _ingest-pipelines/processors/join.md [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+`on_failure` | Optional | Handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:36]:
```suggestion
`on_failure` | Optional | Specifies to handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

File: _ingest-pipelines/processors/join.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+`tag` | Optional | Identifier for the processor. Useful for debugging and metrics.
```
Comment (natebower) [inline _ingest-pipelines/processors/join.md:37]:
```suggestion
`tag` | Optional | An identifier for the processor. Useful for debugging and metrics.
```

**Review by natebower — COMMENTED** [review]
@vagimeli Please see my comments and changes and let me know if you have any questions. Thanks!

**Pull request merged**